modules/renter/worker.go

```diff
@@ -0,0 +1,245 @@
+	// channel for returning the results.
+	uploadWork struct {
+		// data is the payload of the upload.
+		chunkIndex uint64
+		data       []byte
+		file       *file
+		pieceIndex uint64
+
+		// resultChan is a channel that the worker will use to return the
+		// results of the upload.
+		resultChan chan finishedUpload
+	}
+
+	// A worker listens for work on a certain host.
+	worker struct {
+		// contractID specifies which contract the worker specifically works
+		// with.
+		contractID types.FileContractID
+
+		// If there is work on all three channels, the worker will first do all
+		// of the work in the download chan, then all of the work in the
+		// priority upload chan, and finally all of the work in the upload
+		// chan.
+		//
+		// A busy higher priority channel is able to entriely starve all of the
+		// channels with lower priority.
+		downloadChan         chan downloadWork // higher priority than all uploads
+		killChan             chan struct{}     // highest priority
+		priorityDownloadChan chan downloadWork // higher priority than downloads (used for user-initiated downloads)
+		uploadChan           chan uploadWork   // lowest priority
+
+		// recentUploadFailure documents the most recent time that an upload
+		// has failed. The repair loop ignores workers that have had an upload
+		// failure in the past two hours.
+		recentUploadFailure time.Time
+
+		// consecutiveDownloadFailures records the number of download failures
+		// in a row that have been experienced by this worker. This number is
+		// used to determine whether the worker should be removed from the
+		// download pool or not.
+		consecutiveDownloadFailures int
+
+		// Utilities
+		renter *Renter
+	}
+)
+
+// download will perform some download work.
+func (w *worker) download(dw downloadWork) {
+	d, err := w.renter.hostContractor.Downloader(w.contractID)
+	if err != nil {
+		dw.resultChan <- finishedDownload{dw.chunkIndex, nil, err, w.contractID}
+		return
+	}
+	defer d.Close()
+
+	data, err := d.Sector(dw.dataRoot)
+	dw.resultChan <- finishedDownload{dw.chunkIndex, data, err, w.contractID}
+}
+
+// upload will perform some upload work.
+func (w *worker) upload(uw uploadWork) {
+	e, err := w.renter.hostContractor.Editor(w.contractID)
+	if err != nil {
+		uw.resultChan <- finishedUpload{crypto.Hash{}, err, w.contractID}
+		return
+	}
+	defer e.Close()
+
+	root, err := e.Upload(uw.data)
+	if err != nil {
+		uw.resultChan <- finishedUpload{root, err, w.contractID}
+		return
+	}
+
+	// Update the renter metadata.
+	uw.file.mu.Lock()
+	contract, exists := uw.file.contracts[w.contractID]
+	if !exists {
+		contract = fileContract{
+			ID:          w.contractID,
+			IP:          e.Address(),
+			WindowStart: e.EndHeight(),
+		}
+	}
+	contract.Pieces = append(contract.Pieces, pieceData{
+		Chunk:      uw.chunkIndex,
+		Piece:      uw.pieceIndex,
+		MerkleRoot: root,
+	})
+	uw.file.contracts[w.contractID] = contract
+	uw.file.mu.Unlock()
+	w.renter.saveFile(uw.file)
```

Comment:
I felt it was a lot easier to handle here. The repair loop itself receives the completed jobs all down the same channel, and so doesn't have a whole lot of context about what exactly the change was. It complicates the repair loop a lot to have to think about more than just figuring out how to reassign workers.

I will probably move the download / file write code into the worker as well. It'll keep the repair loop and download loop focused on assigning work, instead of handling completed jobs by... completing them more? I feel like part of the job of the worker is to update all the persist, etc.